File: src/main/kotlin/sunsetsatellite/lang/sunlite/Expr.kt
package sunsetsatellite.lang.sunlite

import com.sun.org.apache.xpath.internal.operations.Bool

abstract class Expr: Element {
	data class Binary(val left: Expr, val operator: Token, val right: Expr) : Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitBinaryExpr(this)
		}

		override fun getLine(): Int {
			return operator.line
		}

		override fun getFile(): String? {
			return operator.file
		}

		override fun getExprType(): Type {
            if (operator.type == TokenType.STAR || operator.type == TokenType.SLASH || operator.type == TokenType.MINUS) {
                return Type.NUMBER
            }
			if(operator.type != TokenType.PLUS) {
				return Type.BOOLEAN
			} else {
				if(left.getExprType() == Type.NUMBER && right.getExprType() == Type.NUMBER) {
					return Type.NUMBER
				} else {
					return Type.STRING
				}
			}
        }
	}

	data class Grouping(val expression: Expr) : Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitGroupingExpr(this)
		}

		override fun getLine(): Int {
			return expression.getLine()
		}

		override fun getFile(): String? {
			return expression.getFile()
		}

		override fun getExprType(): Type {
			return expression.getExprType()
		}
	}

	data class Unary(val operator: Token, val right: Expr) : Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitUnaryExpr(this)
		}

		override fun getLine(): Int {
			return operator.line
		}

		override fun getFile(): String? {
			return operator.file
		}

		override fun getExprType(): Type {
			if(operator.type == TokenType.BANG) {
				return Type.BOOLEAN
			} else if(operator.type == TokenType.MINUS) {
				return Type.NUMBER
			}
			return Type.UNKNOWN
		}
	}

	data class Literal(val value: Any?, val lineNumber: Int, val currentFile: String?, val type: Type = Type.UNKNOWN): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitLiteralExpr(this)
		}

		override fun getLine(): Int {
			return lineNumber
		}

		override fun getFile(): String? {
			return currentFile
		}

		override fun getExprType(): Type {
			return type
		}
	}

	data class Variable(val name: Token, val type: Type = Type.UNKNOWN, val constant: Boolean = false): Expr(), NamedExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitVariableExpr(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}

		override fun getExprType(): Type {
			return type
		}
	}

	/*class GenericVariable(name: Token, val typeParameters: List<Param>): Variable(name) {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitVariableExpr(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}
	}*/

	data class Assign(val name: Token, val value: Expr, val operator: TokenType, val type: Type): Expr(), NamedExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitAssignExpr(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}

		override fun getExprType(): Type {
			return type
		}

	}

	data class Logical(val left: Expr, val operator: Token, val right: Expr): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitLogicalExpr(this)
		}

		override fun getLine(): Int {
			return operator.line
		}

		override fun getFile(): String? {
			return operator.file
		}

		override fun getExprType(): Type {
			return Type.BOOLEAN
		}
	}

	data class Call(val callee: Expr, val paren: Token, val arguments: List<Expr>, val typeParams: List<Type>): Expr(), GenericExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitCallExpr(this)
		}

		override fun getLine(): Int {
			return paren.line
		}

		override fun getFile(): String? {
			return paren.file
		}

		override fun getExprType(): Type {
			val type = callee.getExprType()
			if(type is Type.Reference){
				if(type.type == PrimitiveType.CLASS){
					if(!typeParams.isEmpty()){
						val rawType = type.returnType
						return Type.ofGenericObject(rawType.getName(), typeParams)
					}
				}
				return type.returnType
			}
			return Type.UNKNOWN
        }

		override fun getTypeArguments(): List<Type> {
			return typeParams
		}
	}

	data class Lambda(val function: Stmt.Function): Expr(), NamedExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitLambdaExpr(this)
		}

		override fun getLine(): Int {
			return function.name.line
		}

		override fun getFile(): String? {
			return function.getFile()
		}

		override fun getNameToken(): Token {
			return function.name
		}

		override fun getExprType(): Type {
			return Type.ofFunction(function.name.lexeme, function.returnType, function.params)
		}
	}

	data class Get(val obj: Expr, val name: Token, val type: Type = Type.UNKNOWN, val constant: Boolean = false,
				   var typeParams: List<Type> =
					   if(obj.getExprType() is Type.Reference) {
						   val ref = obj.getExprType() as Type.Reference
						   if(ref.type == PrimitiveType.OBJECT){
							   ref.typeParams
						   } else {
							   listOf()
						   }
						} else {
							listOf()
					   })
		: Expr(), NamedExpr, GenericExpr {


		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitGetExpr(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}

		override fun getExprType(): Type {
			return type
		}

		override fun getTypeArguments(): List<Type> {
			return typeParams
		}
	}

	data class ArrayGet(val obj: Expr, val what: Expr, val token: Token, val type: Type = Type.UNKNOWN): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitArrayGetExpr(this)
		}

		override fun getLine(): Int {
			return token.line
		}

		override fun getFile(): String? {
			return token.file
		}

		override fun getExprType(): Type {
			//todo
			return Type.NULLABLE_ANY
		}
	}

	data class Set(val obj: Expr, val name: Token, val value: Expr, val operator: TokenType, val type: Type = Type.UNKNOWN): Expr(), NamedExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitSetExpr(this)
		}

		override fun getLine(): Int {
			return name.line
		}

		override fun getFile(): String? {
			return name.file
		}

		override fun getNameToken(): Token {
			return name
		}

		override fun getExprType(): Type {
			return type
		}

	}

	data class ArraySet(val obj: Expr, val what: Expr, val value: Expr, val token: Token, val operator: TokenType, val type: Type = Type.UNKNOWN): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitArraySetExpr(this)
		}

		override fun getLine(): Int {
			return token.line
		}

		override fun getFile(): String? {
			return token.file
		}

		override fun getExprType(): Type {
			return type
		}
	}

	data class This(val keyword: Token, val type: Type = Type.UNKNOWN): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitThisExpr(this)
		}

		override fun getLine(): Int {
			return keyword.line
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getExprType(): Type {
			return type
		}
	}

	data class Super(val keyword: Token, val method: Token, val type: Type = Type.UNKNOWN): Expr(), NamedExpr {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitSuperExpr(this)
		}

		override fun getLine(): Int {
			return keyword.line
		}

		override fun getFile(): String? {
			return keyword.file
		}

		override fun getNameToken(): Token {
			return method
		}

		override fun getExprType(): Type {
			return type
		}
	}

	data class Check(val left: Expr, val operator: Token, val right: Type): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitCheckExpr(this)
		}

		override fun getLine(): Int {
			return operator.line
		}

		override fun getFile(): String? {
			return operator.file
		}

		override fun getExprType(): Type {
			return Type.BOOLEAN
		}
	}

	data class Cast(val left: Expr, val operator: Token, val right: Type): Expr() {
		override fun <R> accept(visitor: Visitor<R>): R? {
			return visitor.visitCastExpr(this)
		}

		override fun getLine(): Int {
			return operator.line
		}

		override fun getFile(): String? {
			return operator.file
		}

		override fun getExprType(): Type {
			return right
		}
	}

	interface Visitor<R> {
		fun visitBinaryExpr(expr: Binary): R
		fun visitGroupingExpr(expr: Grouping): R
		fun visitUnaryExpr(expr: Unary): R
		fun visitLiteralExpr(expr: Literal): R
		fun visitVariableExpr(expr: Variable): R
		fun visitAssignExpr(expr: Assign): R
		fun visitLogicalExpr(expr: Logical): R
		fun visitCallExpr(expr: Call): R?
		fun visitLambdaExpr(expr: Lambda): R
		fun visitGetExpr(expr: Get): R
		fun visitArrayGetExpr(expr: ArrayGet): R
		fun visitArraySetExpr(expr: ArraySet): R
		fun visitSetExpr(expr: Set): R
		fun visitThisExpr(expr: This): R
		fun visitSuperExpr(expr: Super): R
		fun visitCheckExpr(expr: Check): R
		fun visitCastExpr(expr: Cast): R
	}

	interface NamedExpr: Element {
		fun getNameToken(): Token
	}

	interface GenericExpr: Element {
		fun getTypeArguments(): List<Type>
	}

	abstract fun <R> accept(visitor: Visitor<R>): R?
	abstract fun getExprType(): Type
}

File: src/main/csharp/SunliteSharp/Core/AST/Expr.cs
using SunliteSharp.Core.Enum;

namespace SunliteSharp.Core.AST;

public abstract record Expr : Element
{
    public abstract int GetLine();
    public abstract string GetFile();
    public abstract Type GetExprType();
    
    public interface INamedExpr : Element
    {
        Token GetNameToken();
    }
    
    public interface IGenericExpr : Element
    {
        List<Param> GetTypeArguments();
    }

    public record Binary(Expr Left, Token Operator, Expr Right) : Expr {
        public override int GetLine()
        {
            return Operator.Line;
        }

        public override string GetFile()
        {
            return Operator.File;
        }

        public override Type GetExprType()
        {
            if (Operator.Type is TokenType.Star or TokenType.Slash or TokenType.Minus)
            {
                return Type.Number;
            }

            if (Operator.Type is not TokenType.Plus)
            {
                return Type.Boolean;
            }

            if(Left.GetExprType() == Type.Number && Right.GetExprType() == Type.Number)
            {
                return Type.Number;
            }

            return Type.String;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitBinaryExpr(this);
        }
    }

    public record Grouping(Expr Expression) : Expr
    {
        public override int GetLine()
        {
            return Expression.GetLine();
        }

        public override string GetFile()
        {
            return Expression.GetFile();
        }

        public override Type GetExprType()
        {
            return Expression.GetExprType();
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitGroupingExpr(this);
        }
    }

    public record Unary(Token Operator, Expr Right) : Expr
    {
        public override int GetLine()
        {
            return Operator.Line;
        }
        public override string GetFile()
        {
            return Operator.File;
        }
        public override Type GetExprType()
        {
            return Right.GetExprType();
        }
        public override void Accept(IVisitor visitor)
        {
            visitor.VisitUnaryExpr(this);
        }
    }

    public record Literal(object? Value, int LineNumber, string CurrentFile, Type Type) : Expr
    {
        public override int GetLine()
        {
            return LineNumber;
        }
        public override string GetFile()
        {
            return CurrentFile;
        }
        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitLiteralExpr(this);
        }
    }

    public record Variable(Token Name, Type Type, bool Constant) : Expr, INamedExpr
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitVariableExpr(this);
        }
    }
    
    public record Assign(Token Name, Expr Value, TokenType Operator, Type Type) : Expr, INamedExpr
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitAssignExpr(this);
        }
    }
    
    public record Logical(Expr Left, Token Operator, Expr Right) : Expr
    {
        public override int GetLine()
        {
            return Operator.Line;
        }

        public override string GetFile()
        {
            return Operator.File;
        }

        public override Type GetExprType()
        {
            return Type.Boolean;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitLogicalExpr(this);
        }
    }

    public record Call(Expr Callee, Token Paren, List<Expr> Arguments, List<Param> TypeArgs) : Expr, IGenericExpr
    {
        public override int GetLine()
        {
            return Paren.Line;
        }

        public override string GetFile()
        {
            return Paren.File;
        }

        public List<Param> GetTypeArguments()
        {
            return TypeArgs;
        }

        public override Type GetExprType()
        {
            var type = Callee.GetExprType();
            if (type is not Type.Reference refType) return Type.Unknown;
            if (refType.Type != PrimitiveType.Class || TypeArgs.Count == 0) return refType.ReturnType;
            var rawType = refType.ReturnType;
            return Type.OfGenericObject(rawType.Name(), TypeArgs);
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitCallExpr(this);
        }
    }

    public record Lambda(Stmt.Function Function) : Expr, INamedExpr
    {
        public override int GetLine()
        {
            return Function.Name.Line;
        }

        public override string GetFile()
        {
            return Function.Name.File;
        }

        public Token GetNameToken()
        {
            return Function.Name;
        }

        public override Type GetExprType()
        {
            return Type.OfFunction(Function.Name.Lexeme, Function.ReturnType, Function.Params);
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitLambdaExpr(this);
        }
    }

    public record Get : Expr, INamedExpr, IGenericExpr
    {

        public Get(Expr obj, Token name, Type type, bool constant, List<Param> typeParams)
        {
            Obj = obj;
            Name = name;
            Type = type;
            Constant = constant;
            List<Param> list = [];
            if (obj.GetExprType() is Type.Reference { Type: PrimitiveType.Object } refType)
            {
                list = refType.TypeParams;
            }
            TypeParams = list;
        }

        public Expr Obj;
        public Token Name;
        public Type Type;
        public bool Constant;
        public List<Param> TypeParams;
        
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public List<Param> GetTypeArguments()
        {
            return TypeParams;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitGetExpr(this);
        }
    }

    public record ArrayGet(Expr Obj, Expr What, Token Token) : Expr
    {
        public override int GetLine()
        {
            return Token.Line;
        }

        public override string GetFile()
        {
            return Token.File;
        }

        public override Type GetExprType()
        {
            if (Obj.GetExprType() is Type.Reference { Type: PrimitiveType.Array or PrimitiveType.Table} refType)
            {
                return refType.ReturnType;
            }
            return Type.Unknown;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitArrayGetExpr(this);
        }
    }
    
    public record Set(Expr Obj, Token Name, Expr Value, TokenType Operator, Type Type) : Expr, INamedExpr
    {
        public override int GetLine()
        {
            return Name.Line;
        }

        public override string GetFile()
        {
            return Name.File;
        }

        public Token GetNameToken()
        {
            return Name;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitSetExpr(this);
        }
    }
    
    public record ArraySet(Expr Obj, Expr What, Expr Value, Token Token, TokenType Operator, Type Type) : Expr
    {
        public override int GetLine()
        {
            return Token.Line;
        }

        public override string GetFile()
        {
            return Token.File;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitArraySetExpr(this);
        }
    }

    public record This(Token Keyword, Type Type) : Expr
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitThisExpr(this);
        }
    }

    public record Super(Token Keyword, Token Method, Type Type) : Expr, INamedExpr
    {
        public override int GetLine()
        {
            return Keyword.Line;
        }

        public override string GetFile()
        {
            return Keyword.File;
        }

        public Token GetNameToken()
        {
            return Method;
        }

        public override Type GetExprType()
        {
            return Type;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitSuperExpr(this);
        }
    }

    public record Check(Expr Left, Token Operator, Type Right) : Expr
    {
        public override int GetLine()
        {
            return Operator.Line;
        }

        public override string GetFile()
        {
            return Operator.File;
        }

        public override Type GetExprType()
        {
            return Type.Boolean;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitCheckExpr(this);
        }
    }

    public record Cast(Expr Left, Token Operator, Type Right) : Expr
    {
        public override int GetLine()
        {
            return Operator.Line;
        }

        public override string GetFile()
        {
            return Operator.File;
        }

        public override Type GetExprType()
        {
            return Right;
        }

        public override void Accept(IVisitor visitor)
        {
            visitor.VisitCastExpr(this);
        }
    }
    
    public interface IVisitor
    {
        void VisitBinaryExpr(Binary expr);
        void VisitGroupingExpr(Grouping grouping);
        void VisitUnaryExpr(Unary unary);
        void VisitLiteralExpr(Literal expr);
        void VisitVariableExpr(Variable expr);
        void VisitAssignExpr(Assign expr);
        void VisitLogicalExpr(Logical expr);
        void VisitCallExpr(Call expr);
        void VisitLambdaExpr(Lambda expr);
        void VisitGetExpr(Get expr);
        void VisitArrayGetExpr(ArrayGet expr);
        void VisitArraySetExpr(ArraySet expr);
        void VisitSetExpr(Set expr);
        void VisitThisExpr(This expr);
        void VisitSuperExpr(Super expr);
        void VisitCheckExpr(Check expr);
        void VisitCastExpr(Cast expr);
    }
    
    public abstract void Accept(IVisitor visitor);
    
}